Refactor Constructors page (#5757)
dart-lang/site-www: src/content/language/constructors.md

  title: Methods
---

Constructors create instances of classes.

* To declare a constructor, create a function named the same as its class.
  You can add an optional additional identifier as described in
  [Named constructors](#named-constructors).

* To instantiate a class, use the
  [generative constructor](#generative-constructors).

* To instantiate any instance variables,

>   [initialize formal parameters](#parameter-initialization).

Author: Refactored with explanatory list.

<?code-excerpt path-base="misc/lib/language_tour/classes"?>

<?code-excerpt replace="/ *\/\/\s+ignore_for_file:[^\n]+\n//g; /(^|\n) *\/\/\s+ignore:[^\n]+\n/$1/g; /(\n[^\n]+) *\/\/\s+ignore:[^\n]+\n/$1\n/g; / *\/\/\s+ignore:[^\n]+//g; /([A-Z]\w*)\d\b/$1/g"?>

## Types of constructors

> ### Constructor inheritance

Author: Can do, though I think this has info relevant to the succeeding descriptions of types of constructors. Let's see how it reads.

_Subclasses_, or child classes, don't inherit *constructors*
from their _superclass_, or immediate parent class.

> If a class doesn't declare a constructor, it can only use the
> [default constructor](#default-constructors).

Reviewer: [nit] I don't think this sentence is necessary here.

```suggestion
```

A class can inherit the _parameters_ of a superclass.
These are called [super parameters](#super-parameters)
A subclass without constructor declarations can only use
a [default constructor](#default-constructors).

Constructors work in a somewhat similar way to
how you call a chain of static methods.
Each subclass can call its superclass's constructor to initialize an instance,
like a subclass can call a superclass's static method.

> This process doesn't involve any "inheritance".

Reviewer: [optional] I think it might be more clear to add some specificity.

```suggestion
This process doesn't involve any "inheritance" of constructor bodies or signatures.
```

### Default constructors

If you don't declare a constructor, Dart uses the default constructor.

> The default constructor has no arguments and no name.

Reviewer: [optional] I don't know for sure if the detail of "generative constructor" is specifically useful to understand the concept of the default constructor - but knowing that the default constructor is a generative constructor helps to understand the concept of generative constructors in my opinion. I think since this flows into the next section, it could be worth adding this detail. What do you think?

```suggestion
The default constructor is a generative constructor with no arguments and no name.
```

### Generative constructors

> To instantiate a class, use the generative constructor.

Reviewer: [nit] It's possible for a class to have more than one generative constructor.

```suggestion
To instantiate a class, use a generative constructor.
```

<?code-excerpt "point_alt.dart (idiomatic-constructor)" plaster="none"?>
```dart
class Point {
  double x = 0;

>   double y = 0;

Reviewer: [nit] I don't think it is particularly useful to show these initializing expressions (nor making the fields mutable). I'd keep these two lines as they were.

}
```

A subclass doesn't inherit a superclass's named constructor.
To create a subclass with a named constructor defined in the superclass,
implement that constructor in the subclass.

### Non-default superclass constructors

Dart executes constructors in the following order:

1. [initializer list](#use-an-initializer-list)
1. superclass's unnamed, no-arg constructor

> 1. main class's no-arg constructor

Reviewer: This bullet point feels out of place - that's not a new problem in this PR though.

I don't know why we need to talk specifically about the arguments to the main class constructor - this description applies to any constructor definition that doesn't have a super constructor invocation.

I was trying to find a suggestion for a rewrite of this section and it struck me that waiting until the end of the doc to show what an "initializer list" is may be hurting us.

I wonder if we should try to show syntax and terminology near the top of this doc. If a reader knows that the "initializer list" is "the part with `fieldName = parameterName` because they saw that syntax labeled, I think they would have an easier time understanding what it means for the "initializer list" to run in a given position in the instantiation process, even if they don't know the details that will be introduced later.

Author: Understood. I changed the first example to call out what the initializer list was.

}
```

To further illustrate, consider the following example.

```dart
  // If you invoke the super constructor (`super(0)`) with any
  // positional arguments, using a super parameter (`super.x`)

>   // returns an error.

Reviewer: [nit] I don't know what "returns" would mean in this context. How about "is an error" or "results in an error"?

```suggestion
  // is an error.
```

}
```

### Constant constructors

If your class produces unchanging objects, make these
objects compile-time constants. 
To make objects compile-time constants, define a `const` constructor
with all instance variables set as `final`.

<?code-excerpt "immutable_point.dart"?>
```dart
class ImmutablePoint {
  static const ImmutablePoint origin = ImmutablePoint(0, 0);

  final double x, y;

  const ImmutablePoint(this.x, this.y);
}
```

> Constant constructors don't always create constants.

Reviewer: [optional]

```suggestion
Constant constructors may be invoked in a non-const context and don't always create constants.
```

}
```

### Factory constructors

When encountering one of two cases of implementing a constructor,

> use the `factory` keyword:

Reviewer: [nit] Is this missing a word at the end? `use the factory keyword when:` ? `use the factory keyword if:` ?

var loggerJson = Logger.fromJson(logMap);
```

### Redirecting factory constructors

A redirecting factory constructor specifies a call to a constructor of another
class to use whenever someone makes a call to the redirecting constructor.

```dart
factory Listenable.merge(List<Listenable> listenables) = _MergingListenable
```

It might appear that ordinary factory constructors
could create and return instances of other classes.
This would make redirecting factories unnecessary.
Redirecting factories have several advantages:

* An abstract class might provide a constant constructor
  that uses the constant constructor of another class.
* A redirecting factory constructor avoids the need for forwarders
  to repeat the formal parameters and their default values.

## Parameter initialization

> Dart can initialize parameters in a constructor in three ways.

Author: Fixed.

  }
}
```

### Use initializing formal parameters

To simplify the common pattern of assigning a constructor argument
to an instance variable, Dart has *initializing formal parameters*.

In the constructor declaration, include `this.<propertyName>`
and omit the body. The `this` keyword refers to the current instance.

When the name conflict exists, use `this`.
Otherwise, Dart style omits the `this`.
An exception exists for the generative constructor where
you must prefix the initializing formal parameter name with `this`.

As noted earlier in this guide, certain constructors
and certain parts of constructors can't access `this`. These include:

* Factory constructors
* The right-hand side of an initializer list
* Arguments to a superclass constructor

Initializing formal parameters also allow you to initialize
non-nullable or `final` instance variables.
Both of these types of variables require initialization or a default value.

<?code-excerpt "point_alt.dart (initialize-formal)" plaster="none"?>
```dart
class PointB {
  final double x;
  final double y;

  // Sets the x and y instance variables
  // before the constructor body runs.
  PointB(this.x, this.y);

>   // Initializing formal parameters can also be optional.
>   PointB.optional([this.x = 0.0, this.y = 0.0]);
>   PointB.named({required this.x, required this.y});
> }

Reviewer: With the current layout of the doc the last line is covered well in `PointC`- I think it flows well now to omit the mention here. The optional positional parameters are still useful to show here.

```suggestion
  // Initializing formal parameters can also be optional.
  PointB.optional([this.x = 0.0, this.y = 0.0]);
}
```

Author: Fair, but then we should "name" `PointC` constructors to illustrate the point.

```

Private fields can't be used as named initializing formals.

{% comment %}
Don't attach the following example to a code excerpt.
It doesn't work on purpose and will cause errors in CI.
{% endcomment %}
```dart
class PointB {
// ...

  PointB.namedPrivate({required double x, required double y})
      : _x = x,
        _y = y;

// ...
}
```

> This also works with named variables.

Reviewer: I think we should mention named variables before private fields. WDYT about flipping them - move lines 426-442 to after the example currently ending on line 464

Author: Not sure. Let me commit these changes first and you can have another look.
